Frameworks/Foundation/NSProxy.mm

```diff
@@ -14,220 +14,317 @@
  @Notes
 */
 - (id)performSelector:(SEL)aSelector withObject:(id)anObject withObject:(id)anotherObject {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:[self methodSignatureForSelector:aSelector]];
+    invocation.selector = aSelector;
+    [invocation setArgument:anObject atIndex:2];
+    [invocation setArgument:anotherObject atIndex:3];
+    [self forwardInvocation:invocation];
+
+    id ret;
+    [invocation getReturnValue:&ret];
+    return ret;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)isProxy {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return YES;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (instancetype)retain {
-    UNIMPLEMENTED();
-    return StubReturn();
+    intptr_t* refCount = ((intptr_t*)self) - 1;
+    // Note: this should be an atomic read, so that a sufficiently clever
+    // compiler doesn't notice that there's no happens-before relationship
+    // here.
+    if (*refCount >= 0) {
+        __sync_add_and_fetch(refCount, 1);
+    }
+    return self;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (oneway void)release {
-    UNIMPLEMENTED();
+    intptr_t* refCount = ((intptr_t*)self) - 1;
+    // We allow refcounts to run into the negative, but should only
+    // deallocate once.
+    if (__sync_sub_and_fetch(refCount, 1) == -1) {
+        objc_delete_weak_refs(self);
+        [self dealloc];
+    }
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (instancetype)autorelease {
-    UNIMPLEMENTED();
-    return StubReturn();
+    [NSAutoreleasePool addObject:self];
+    return self;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (NSUInteger)retainCount {
-    UNIMPLEMENTED();
-    return StubReturn();
+    // account for the implicit retain on object creation
+    return *(((intptr_t*)self) - 1) + 1;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
+ @Notes Zones are not supported.
 */
 - (NSZone*)zone {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return nil;
+}
+
+#pragma endregion
+
+/**
+ *  _ARCCompliantRetainRelease signals to the runtime (libobjc2) that we are ARC-compliant.
+ *
+ *  When an ARC-compliant class is used with ARC code, it falls into the libobjc2 fast path:
+ *  the runtime automatically increments the refcount without consulting the object, and
+ *  calls dealloc when it crosses below 0.
+ *
+ *  The fast path is only engaged if retain, autorelease, and release are implemented at
+ *  the same inheritance level as _ARCCompliantRetainRelease. Subclasses overriding this behaviour
+ *  opt out of the fast path, and must implement both retain and release.
+ *
+ *  To actually be ARC compliant, we need to store our refcount sizeof(intptr_t) bytes before
+ *  self->isa, and mimic the runtime's behaviour for direct invocations of retain/release.
+ */
+- (void)_ARCCompliantRetainRelease {
 }
 
-@end
+@end
```

Comment:
you may have to implement doesNotRecognizeSelector:, and call it from the default forwardInvocation: implementation just like NSObject -- that's how people will get missing selector exceptions